networkx/generators/directed.py

```diff
@@ -485,17 +489,64 @@ def random_k_out_graph(n, k, alpha, self_loops=True, seed=None):
     """
     if alpha < 0:
         raise ValueError("alpha must be positive")
+    return _random_k_out_graph_numpy(n, k, alpha, self_loops, seed)
+
+
+@np_random_state(4)
+def _random_k_out_graph_numpy(n, k, alpha, self_loops=True, seed=None):
+    import numpy as np
+
+    G = nx.empty_graph(n, create_using=nx.MultiDiGraph)
+    nodes = np.arange(n)
+    remaining_mask = np.full(n, True)
+    weights = np.full(n, alpha)
+    total_weight = n * alpha
+    out_strengths = np.zeros(n)
+
+    for i in range(k * n):
+        u = seed.choice(nodes[remaining_mask])
+
+        if not self_loops:
+            u_weight = weights[u]
+            weights[u] = 0
+            total_weight -= u_weight
+
+        v = seed.choice(nodes, p=weights / total_weight)
+
+        if not self_loops:
+            weights[u] = u_weight
+            total_weight += u_weight
```

Comment:
I was initially thrown by the pattern of "conditionally set state, do calculation, conditionally reset state" here, but it's clear once I took the time to read it! An alternative might be:

```suggestion
        if self_loops:                                                          
            v = seed.choice(nodes, p=weights / total_weight)                    
        else:  # Ignore weight of u when selecting v                            
            u_weight = weights[u]                                               
            weights[u] = 0                                                      
            v = seed.choice(nodes, p=weights / (total_weight - u_weight))       
            weights[u] = u_weight
```

This is purely subjective, so feel free to ignore - it's fine as-is!

Reply:
I agree this reads better, if only for the fact that my implementation was defining the variable `u_weight` only within a conditional branch, and then used it in an another branch (although based on the same condition)